### PR TITLE
openapi 3.0.2

### PR DESCRIPTION
**Pull request by iz-podpolja**

### CODE REVIEW

File: betterdocs.gemspec [hunk 1]
```diff
@@ -1,49 +1,49 @@
-# -*- encoding: utf-8 -*-
 # stub: betterdocs 0.7.1 ruby lib
 
 Gem::Specification.new do |s|
-  s.name = "betterdocs".freeze
```
Comment (jaynetics) [inline betterdocs.gemspec:5]: this file is auto-generated so it doesn't make sense to format it

Comment (iz-podpolja) [reply betterdocs.gemspec:5]: Probably did Cmd+Shift+S when it was open, I'll see if there is a .solargraphignore or to this effect

File: lib/betterdocs/generator/swagger.rb [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+      def get_param_value_type(value)
+        return 'string' if value.instance_of? String
+        return 'integer' if value.instance_of? Integer
+        return 'boolean' if value.instance_of? TrueClass
+        return 'boolean' if value.instance_of? FalseClass
+        return 'number' if value.instance_of? Float
+        return 'array' if value.instance_of? Array
+
+        'object'
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:123]:
```suggestion
        case value
        when String then 'string'
        when Integer then 'integer'
        when TrueClass, FalseClass then 'boolean'
        when Float then 'number'
        when Array then 'array'
        else 'object'
        end
```

a bit more idiomatic

File: lib/betterdocs/generator/swagger.rb [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+            cmp = "{#{name}}#{suffix}"
+            param = { name: name, type: get_path_param_slug_type(split[i].gsub('.json', '')) }
+            params.push(param)
+          end
+          path_cmp.push(cmp)
+        end
+        slugs[:path] = "/#{path_cmp.join('/')}"
+        slugs[:params] = params
+        slugs
+      end
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:179]:
```suggestion
      def get_request_url_slugs(request)
        slugs = %r{
          (?<method>GET|POST|PUT|DELETE|OPTIONS)\s+
          (?<protocol>https?)://
          (?<host>[a-z.]+)/
          (?<lang>[a-z]{2})/api_
          (?<ver>v[0-9]+)/
          (?<path>[\w/]+\.json?)
        }x.match(request).named_captures.transform_keys(&:to_sym)
        slugs[:params] = []
        path_cmp = slugs[:path].split('/').map.with_index do |cmp, i|
          unless i.even?
            suffix = cmp['.json'] || ''
            name = "#{split[i - 1]}_id"
            cmp = "{#{name}}#{suffix}"
            param = { name: name, type: get_path_param_slug_type(cmp.gsub('.json', '')) }
            slugs[:params].push(param)
          end
          cmp
        end
        slugs[:path] = "/#{path_cmp.join('/')}"
        slugs
      end
```

File: lib/betterdocs/generator/swagger.rb [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        if action.json_params.full?
+          payload_name = "#{name.titleize.gsub(/\s/, '')}Request"
+          definitions[payload_name] = get_request_definition(action.json_params.values)
+          params[:requestBody] = wrap_content_object({ schema: get_schema_ref(payload_name) }, "#{name} Request")
+        end
+      end
+
+      def add_path_params(params, from_path)
+        from_path.each do |param|
+          params = add_param(params, param[:name], 'path', true, param[:type], param[:name])
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:313]:
```suggestion
          add_param(params, param[:name], 'path', true, param[:type], param[:name])
```
assignment is unused

File: lib/betterdocs/generator/swagger.rb [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+        end
+      end
+
+      def add_query_params(params, from_query, is_a_list)
+        from_query.each do |param|
+          params = add_param(params, param.name, 'query', param.required, get_param_value_type(param.value),
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:319]:
```suggestion
          add_param(params, param.name, 'query', param.required, get_param_value_type(param.value),
```

File: lib/betterdocs/generator/swagger.rb [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+        }
+      end
+
+      def add_required_to_definitions(definitions)
+        definitions.each do |def_key, d|
+          req = []
+          d[:properties].each do |key, p|
+            req.push(key) if p.key?(:nullable)
+          end
+          definitions[def_key][:required] = req unless req.length.zero?
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:299]:
```suggestion
          req = d[:properties].select { |k, v| v.key?(:nullable) }.keys
          definitions[def_key][:required] = req unless req.empty?
```

File: lib/betterdocs/generator/swagger.rb [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+        end
+        add_default_list_params(params) if is_a_list
+      end
+
+      def wrap_content_object(object, description)
+        { content: { "application/json": object }, description: description }
+      end
+
+      def add_example(obj, name, example)
+        obj[:examples] = {} unless obj.key?(:examples)
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:330]:
```suggestion
        obj[:examples] ||= {}
```

File: lib/betterdocs/generator/swagger.rb [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+        message = color(
+          231,
+          on_color(124, " *** ERROR #{exception.class}: #{exception.message.inspect} ***")
+        )
+        infobar.puts message
+        exit 1
+      end
+
+      def add_param(hash, name, param_in, required = false, type = nil, description = '', schema = nil)
+        hash[:parameters] = [] unless hash.key?(:parameters)
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:92]:
```suggestion
        hash[:parameters] ||= []
```

File: lib/betterdocs/generator/swagger.rb [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+      end
+
+      def get_list_response_wrapper_schema
+        { type: 'object', properties: {
+          total_entries: { type: 'integer' },
+          offset: { type: 'integer' },
+          total_pages: { type: 'integer' },
+          current_page: { type: 'integer' },
+          per_page: { type: 'integer' }
+        }, required: %w[total_entries offset current_page per_page total_pages data] }
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:112]:
```suggestion
        { 
          type: 'object',
          properties: {
            total_entries: { type: 'integer' },
            offset: { type: 'integer' },
            total_pages: { type: 'integer' },
            current_page: { type: 'integer' },
            per_page: { type: 'integer' }
          },
          required: %w[total_entries offset current_page per_page total_pages data]
        }
```

more standard formatting

File: lib/betterdocs/generator/swagger.rb [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+        end
+        res[:nullable] = true if nullable
+        res
+      end
+
+      def get_dto_name(representer)
+        return representer.name.demodulize if representer&.name
+
+        nil
+      end
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:245]:
```suggestion
      def get_dto_name(representer)
        representer.name&.demodulize
      end
```

File: lib/betterdocs/generator/swagger.rb [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+        unless definition[:properties][:links]
+          definition[:properties][:links] = {
+            type: 'array',
+            nullable: false,
+            items: { type: 'object', properties: {
+              rel: { type: 'string', enum: [] },
+              href: { type: 'string' }
+            }, required: %w[rel href] }
+          }
+        end
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:258]:
```suggestion
        definition[:properties][:links] ||= {
          type: 'array',
          nullable: false,
          items: {
            type: 'object',
            properties: {
              rel: { type: 'string', enum: [] },
              href: { type: 'string' }
            },
            required: %w[rel href]
          }
        }
```

File: lib/betterdocs/generator/swagger.rb [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+        enum = definition[:properties][:links][:items][:properties][:rel][:enum]
+        enum.push(value) unless enum.include?(value)
+      end
+
+      def initialise_definition(definitions, cls)
+        definitions[cls] = { type: 'object', properties: {} } if definitions[cls].nil?
+        definitions[cls]
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:265]:
```suggestion
        definitions[cls] ||= { type: 'object', properties: {} }
```

File: lib/betterdocs/generator/swagger.rb [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+      end
+
+      def add_response_schema(definitions, response)
+        return nil if response.full?.nil?
```
Comment (jaynetics) [inline lib/betterdocs/generator/swagger.rb:269]:
```suggestion
        return nil unless response.full?
```

**Pull request merged**